### PR TITLE
fix(subtreevalidation): cap NewIncompleteTreeByLeafCount allocation at policy max

### PR DESCRIPTION
**Pull request by ordishs**
Closes #4572.

## Summary
`NewIncompleteTreeByLeafCount(len(subtreeNodeBytes) / HashSize)` derives its capacity from peer-supplied bytes. Without an explicit cap, a multi-gigabyte response would have caused `make([]Node, 0, leafCount)` to allocate at peer-controlled size — an OOM vector.

## Fix
Inserts an explicit bounds check immediately before `NewIncompleteTreeByLeafCount`:

```go
leafCount := len(subtreeNodeBytes) / chainhash.HashSize
if err := validateSubtreeLeafCount(subtreeHash, leafCount, u.settings.BlockAssembly.MaximumMerkleItemsPerSubtree); err != nil {
    return err
}
```

The check is extracted into a small `validateSubtreeLeafCount` helper so the policy contract is unit-testable in isolation.

## Relationship to #4571
PR #772 (issue #4571, on the same audit branch) bounds the HTTP body read at this call site to `MaximumMerkleItemsPerSubtree * HashSize` bytes — so the OOM vector is also blocked one layer up. This commit adds the explicit bound at the allocation layer for defense-in-depth and matches the audit's prescribed shape.

## Test plan
- [x] Unit test for the bounds check: leaf count over the cap returns ProcessingError; leaf count at or under the cap succeeds.
- [x] `go test -race -tags testtxmetacache ./services/subtreevalidation/...` passes.

### CODE REVIEW

**Review by freemans13 — APPROVED** [review]

File: services/subtreevalidation/check_block_subtrees.go [hunk 1]
```diff
@@ -223,7 +223,12 @@ func (u *Server) CheckBlockSubtrees(ctx context.Context, request *subtreevalidat
 						}
 					}
 
-					subtreeToCheck, err = subtreepkg.NewIncompleteTreeByLeafCount(len(subtreeNodeBytes) / chainhash.HashSize)
+					leafCount := len(subtreeNodeBytes) / chainhash.HashSize
```
Comment (icellan) [inline services/subtreevalidation/check_block_subtrees.go:226]: ✅ Addressed in this PR!

The legacy block ingestion path now uses partitionLegacyBlock (added in services/legacy/netsync/subtree_partition.go) which respects settings.BlockAssembly.MaximumMerkleItemsPerSubtree at line 295 of handle_block.go:

maxItems := sm.settings.BlockAssembly.MaximumMerkleItemsPerSubtree
subtreeSize, numSubtrees, finalLeafCount, err := partitionLegacyBlock(txCount, maxItems)

This ensures legacy blocks are split into multiple subtrees when they exceed the configured maximum, matching the validation rules in model.Block.CheckMerkleRoot.

**Review by icellan — CHANGES_REQUESTED** [review]

**Review by oskarszoon — CHANGES_REQUESTED** [review]
Original audit fix in place and correct — `validateSubtreeLeafCount` caps peer-supplied leaf counts at `MaximumMerkleItemsPerSubtree` before allocation, closing #4572 cleanly. The expanded subtree-lift work is consensus-relevant but well-reasoned: `RootHashPadded` is the right primitive (upstream-compatible addition in go-subtree v1.3.2, not a fork), the math equivalence (flat-merkle ↔ partition+top-tree+lift) is pinned by `util/subtree_merkle_root_equivalence_test.go`, and the audit notes correctly establish that historical Teranode-produced blocks fall in the unchanged 1-subtree branch. Validator + assembly + legacy ingest are updated in lockstep. The new duplicate-subtree-root detection in `CheckMerkleRoot` is a good CVE-2012-2459-style hardening.

Three things blocking a clean approve:

**Gap 1 (gating) — `CheckMerkleRoot` doesn't enforce `IsPowerOfTwo(SubtreeSlices[0].Length())`.** This is the load-bearing precondition for the lift-vs-canonical-flat-merkle equivalence — the math only works when the first subtree's length is a power of two. The check at `model/Block.go:1340-1342` validates non-last subtrees match the first's length and last-subtree-when-incomplete is a power of two, but never validates the first itself. A peer can craft a block with `Subtrees[0].Length = 3, Subtrees[1].Length = 2`; the lift then runs `RootHashPadded(2)` and produces a top-tree root that doesn't match what a canonical SV Node validator would compute from the flat tx list. With **no activation gate / FSM hook / feature flag** in this PR (Strategy A applied unconditionally — confirmed: zero matches for `SubtreeLiftActivationHeight`, `liftActivation`, or any FSM gate), this is a chain-split-of-one risk for any Teranode the attacker targets. Three-line fix:

```go
if !subtreepkg.IsPowerOfTwo(targetLength) {
    return errors.NewBlockInvalidError("[BLOCK][%s] first subtree leaf count is not a power of two: %d", b.String(), targetLength)
}
```

between lines 1340 and 1342.

**Gap 2 (gating) — `createMerkleTreeFromSubtrees` doesn't mirror duplicate-subtree-root detection.** `CheckMerkleRoot:1384-1391` now rejects duplicated subtree roots (CVE-2012-2459 hardening), but the assembly-side mirror at `services/blockassembly/Server.go:1484-1533` doesn't. Producer/validator asymmetry. Unlikely to matter today (assembly controls its own subtree set) but if a future bug in `getIncompleteSubtreeMiningData` ever re-emits the same subtree slice, assembly silently produces a block the validator rejects. Mirror the check for symmetry.

**Review state** — worth flagging before merge: freemans13's APPROVED was on commit 1 of 7 (`ffa5554b`, the small +~30 line audit fix). icellan's CHANGES_REQUESTED was on the same. Both reviews predate the expanded scope by ~1100 lines of consensus-adjacent code (commits 2-7: `CheckMerkleRoot` rewrite, assembly-side mirror, legacy partition algorithm, `RootHashPadded` dep bump). The existing APPROVED state shouldn't be taken as cover for the expanded scope — recommend explicitly asking at least one reviewer who understands the consensus implications to walk commits 2-7 fresh.

Non-gating follow-ups (worth filing but won't block merge):

a. Audit `DoHTTPRequestBodyReader` for content-length / streaming caps. `validateSubtreeLeafCount` fires *after* bytes are in memory — if the HTTP layer doesn't enforce a body-size cap matched to `MaximumMerkleItemsPerSubtree × HashSize`, a peer can still cause a transient memory spike.
b. Explicit allocation cap in `NewBlockFromMsgBlock` (`model/Block.go:154`). Today bounded implicitly by P2P-layer MaxBlockSize.
c. Expand `partitionLegacyBlock` table tests to cover `totalLeaves = maxItems + 1` (smallest-possible-tail, r=1) and `totalLeaves = 2 * maxItems` (K=2 complete, lift no-op).
d. Document the "first subtree must be a power of two" invariant inline at `CheckMerkleRoot` and `createMerkleTreeFromSubtrees` so a future "consolidation" PR doesn't silently break it.
e. Convert `plans/subtree-lift-audit-notes.md` from a "planned change" doc to a permanent design-decision record (the change has now landed; the file should reflect that).

Happy to re-approve once Gaps 1 and 2 are closed and at least one fresh human review covers commits 2-7.

**Review by icellan — APPROVED** [review]

**Review by oskarszoon — APPROVED** [review]
Re-reviewed `75952378f2`. Both gaps closed cleanly:

- Gap 1 (`IsPowerOfTwo(targetLength)` guard in `CheckMerkleRoot`) is placed correctly between `targetLength := SubtreeSlices[0].Length()` and the loop, so the `[3, 2]` attack fails fast with `BlockInvalidError` before any lift work. Edge cases check out (length=0 unreachable in practice, length=1 passes as a power of two).
- Gap 2 (duplicate-subtree-root mirror in `createMerkleTreeFromSubtrees`) matches the validator-side check exactly, with `ProcessingError` on the assembly side — correct, a duplicate here would be a producer-side regression, not peer data.

Tests are proportionate: `TestCheckMerkleRoot_RejectsFirstSubtreeNotPowerOfTwo` builds a real `[3-leaf, 4-leaf]` block fixture and asserts both `errors.Is(err, ErrBlockInvalid)` + substring. `TestCreateMerkleTreeFromSubtrees_RejectsDuplicateSubtreeRoots` includes a `require.Equal(first.RootHash, second.RootHash)` self-validation guarding against future fixture-builder bitrot — nice touch. `TestSubtreeMerkleRootEquivalence` still holds. CI is 25/0 green.

Strictly additive — 4 files, +104/-0, no scope creep.

Approve. The 5 non-gating follow-ups from the prior review (HTTP body-size cap audit, explicit `NewBlockFromMsgBlock` cap, partition edge-case tests, document the pow2 invariant inline, convert audit-notes plan to a permanent design record) still stand whenever you want to pick them up.
